pkg/engine/constraints.go

```diff
@@ -0,0 +1,127 @@
+package engine
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/klothoplatform/klotho/pkg/core"
+	"github.com/klothoplatform/klotho/pkg/graph"
+	"gopkg.in/yaml.v3"
+)
+
+type (
+	Constraint interface {
+		Level() ConstraintLevel
+		IsSatisfied(dag *core.ResourceGraph) bool
+		CanApply() bool
+		Apply(dag *core.ResourceGraph) error
```

Comment:
I think we may need more than just a binary error/no-error. The caller may want to know some details about _how_ the constraint was used, and especially if it wasn't used.

For example, let's say I have a graph `APIGW -> Lambda`, and I say "only use managed storage services." That constraint is a no-op: it's satisfiable by doing nothing, but we should probably provide some way for infracopilot to know that it turned into a no-op constraint, so that we can report that back to the user as a warning.

Reply:
yeah so i think in the engine we need to build a Decision matrix that can show what we did based on constraints